### PR TITLE
Support LLVM 10

### PR DESCRIPTION
**Pull request by thedataking**
Notable changes:
- LLVM 10 sources are hosted on GitHub.
- LLVM 10 uses C++14 features.
- Clang's handling of comments changed. 
- The AST exporter depends on `LLVMFrontendOpenMP` w/LLVM 10.

### CODE REVIEW

File: c2rust-ast-exporter/src/AstExporter.cpp [hunk 1]
```diff
@@ -2417,22 +2421,44 @@ class TranslateConsumer : public clang::ASTConsumer {
...
             for (auto comment : comments) {
                 CborEncoder entry;
                 cbor_encoder_create_array(&array, &entry, 4);
 #if CLANG_VERSION_MAJOR < 8
                 SourceLocation loc = comment->getLocStart();
-#else
+#else // 7 < CLANG_VERSION_MAJOR < 10
                 SourceLocation loc = comment->getBeginLoc();
-#endif // CLANG_VERSION_MAJOR
+#endif // CLANG_VERSION_MAJOR < 10
```
Comment (ahomescu) [inline c2rust-ast-exporter/src/AstExporter.cpp:2435]: Isn't that for the `< 8` check?

File: c2rust-ast-exporter/src/AstExporter.cpp [hunk 2]
```diff
@@ -2417,22 +2421,44 @@ class TranslateConsumer : public clang::ASTConsumer {
...
+                auto raw_text = comment->getRawText(sourceMgr);
                 cbor_encode_byte_string(&entry, raw_text.bytes_begin(),
                                         raw_text.size());
                 cbor_encoder_close_container(&array, &entry);
             }
+#else  // CLANG_VERSION_MAJOR >= 10
+            const FileID file = sourceMgr.getMainFileID();
+            auto comments = Context.getRawCommentList().getCommentsInFile(file);
+            if (comments != nullptr) {
+                cbor_encoder_create_array(&outer, &array, comments->size());
```
Comment (ahomescu) [inline c2rust-ast-exporter/src/AstExporter.cpp:2446]: This looks identical to the code of the other branch, you could probably de-duplicate it using a function or a closure.

File: c2rust-ast-exporter/src/AstExporter.cpp [hunk 3]
```diff
@@ -2467,7 +2493,11 @@ class TranslateAction : public clang::ASTFrontendAction {
     virtual std::unique_ptr<clang::ASTConsumer>
     CreateASTConsumer(clang::CompilerInstance &Compiler,
                       llvm::StringRef InFile) {
-        if(this->getCurrentFileKind().getLanguage() != InputKind::Language::C) {
+#if CLANG_VERSION_MAJOR < 10
+    if(this->getCurrentFileKind().getLanguage() != InputKind::Language::C) {
```
Comment (ahomescu) [inline c2rust-ast-exporter/src/AstExporter.cpp:2497]: You could merge a lot of the two branches using line tricks such that only `InputKind::Language::C` and `Language::C` are under the conditional.

File: c2rust-ast-exporter/src/CMakeLists.txt [hunk 4]
```diff
@@ -49,7 +49,11 @@ if( PROJECT_NAME STREQUAL "LLVM" )
   add_clang_executable(c2rust-ast-exporter ${AST_EXPORTER_BIN_SRCS} DEPENDS clang-headers)
   add_clang_library(clangAstExporter ${AST_EXPORTER_SRCS} DEPENDS clang-headers)
 
-  set(LLVM_LINK_COMPONENTS support)
+  if( PROJECT_VERSION VERSION_LESS "10.0.0" )
+    set(LLVM_LINK_COMPONENTS Support)
+  else()
+    set(LLVM_LINK_COMPONENTS Support FrontendOpenMP)
```
Comment (ahomescu) [inline c2rust-ast-exporter/src/CMakeLists.txt:55]: Could just use `list(APPEND)` to add the library for LLVM 10 and up.

File: c2rust-ast-exporter/src/CMakeLists.txt [hunk 5]
```diff
@@ -87,7 +91,7 @@ endif()
 add_definitions(-DCLANG_LIBDIR_SUFFIX="${LLVM_LIBDIR_SUFFIX}")
 
 set_target_properties(c2rust-ast-exporter PROPERTIES
-  CXX_STANDARD 11
+  CXX_STANDARD 14
```
Comment (ahomescu) [inline c2rust-ast-exporter/src/CMakeLists.txt:94]: Does this work for all older versions, e.g., LLVM 6.0?

File: c2rust-ast-exporter/src/AstExporter.cpp (hunk 2 above)
Comment (thedataking) [reply c2rust-ast-exporter/src/AstExporter.cpp:2446]: It is close but not identical. Example:
`auto raw_text = comment->getRawText(sourceMgr);` vs
`auto raw_text = comment.second->getRawText(sourceMgr);`
I had to use so many preprocessor directives that the code became unreadable. I figured we could live with the duplication until we drop support for anything before LLVM 8 and thus can get rid of a lot of the preprocessor definitions.

**Pull request merged**